### DISCUSSION TITLE
When Controls cleanup happening and how to hook on their "dispose" event?

### DISCUSSION THREAD

**Discussion by SergTomcat** [dotnet/maui#14908] [Q&A] [answered]
When I create custom controls, I may have some events added, or tasks running, or have some instance saved in property that require to be properly disposed.

How can I achieve this?

Example from air: 

```
public class MyView: ContentView
{
......
 protected override OnDestroy() //???
 {
   // do my stuff
    base.OnDestroy();
 }

// or

 public MyView()
 {
// ???
      this.OnDestroying += ()=> { 
     // do my stuff
 }
}


}
```

**Comment by Xyncgas** [comment]
Have you tried implementing IDisposable and write down public void Dispose() in your code and put break point in it see if it's called

**Reply by SergTomcat** [reply]
> Have you tried implementing IDisposable and write down public void Dispose() in your code and put break point in it see if it's called

Tried it, no calling :( 
When I remove parent page from navigation stack, or even when manually removed element from parent ((MyView.Parent as Grid).Remove(MyView)).

```
public class MyView: ContentView, IDisposable
{
                public void Dispose()
		{
			throw new Exception("DISPOSING!"); // no hit
		}
}
```

**Comment by SergTomcat** [comment]
Tried to do it with "behaviour", but "Detatch" is never called when component is recycling.

```
	public class CleaningLadyBehaviour : Behavior<View>
	{
		public delegate void Mop();
		public event Mop MopEvent;

		public CleaningLadyBehaviour(params Mop[] initevents)
		{
			foreach( var evt in initevents)
				MopEvent += evt;
		}

		protected override void OnAttachedTo(View view)
		{
			base.OnAttachedTo(view);
		}

		protected override void OnDetachingFrom(View view)
		{
			foreach (Delegate d in MopEvent?.GetInvocationList())
			{
				var mopEvent = (Mop)d;
				if (mopEvent != null)
				{
					mopEvent.Invoke();
					MopEvent -= mopEvent;
				}
			}


			base.OnDetachingFrom(view);
		}
	}

```

Any help is appreciated.

**Reply by pictos** [reply, marked as answer]
Yeap, from docs you will see the `OnDetach` method isn't called all the time. You can change you `Behavior` to `PlatformBehavior`, the usage is the same but it should call `OnDetach`

**Reply by SergTomcat** [reply]
> Yeap, from docs you will see the `OnDetach` method isn't called all the time. You can change you `Behavior` to `PlatformBehavior`, the usage is the same but it should call `OnDetach`

PlatformBehavior has two methods to override: OnAttachedTo and OnDetatchedFrom, both require FrameworkElement, which is available only for WinUI. I need Droid and iOS also.
Behaviour's methods (OnAttachedTo and OnDetachingFrom) are sealed.

Am I mising something?

**Reply by pictos** [reply]
Yeah, you're. If you don't want/need to use the platform control you can just ignore it... I recorded [this video](https://www.youtube.com/watch?v=khBN1AIL_OM) teaching how to use it and if you can see some code implementation on the [CommunityToolkit.Maui](https://github.com/CommunityToolkit/Maui/tree/main/src/CommunityToolkit.Maui/Behaviors/PlatformBehaviors)

**Reply by SergTomcat** [reply]
I'm sorry, was completey absent-minded and wrote stupid comment. Need to sleep more. Please, ignore it.
Trick with platform view works, OnDetatchingFrom is being called.

If Behavior is the only way to handle component cleanup now, that is the answer.

**Comment by Cybrosys** [comment]
Are there any plans for MAUI to actually call Dispose or add some other method that gets called on `ContentView`? I just encountered this where I had an animation running forever. I figured out I could check if `Window` was null or not to know when my ContentView was no longer being visible/used.

I tried what you suggested above and created a `DisposeBehavior` inheriting from `PlatformBehavior`, seems to work as well.

**Update**
When a page is in a shell tab, the `OnDetachedFrom` is called every time the user leaves the tab; this means that it might not be the best coupling for calling Dispose on something. It is probably better to create specific AttachedTo and DetachedFrom methods on the `ContentView` or just go back to relying on the Window property being set to null.

**Comment by SergTomcat** [comment]
With using _PlatformBehavior_, found some unexpected thing. Behavior detaches even when component is not being recycled, but just lost visibility.
For example, I have Frame on a page, when page is covered by another page (navigation stack, for example), but still exist, behavior on that Frame still detaches. When I come back to the first page and it is appeared again, behavior attaches again.

In my case I want to detect exactly when the control is due for recycling, not in dormant state. Any help?

Upd:
Missed the comment from @Cybrosys above, my bad. That's exactly what I am talking about.

**Comment by BaltoAF** [comment]
Thank you very much. Using PlatformBehavior  is great.
I'm posting because it took me some time to figure out how to use `PlatformBehavior `without worrying about avoiding platform-specific code (e.g., `#if IOS / Android / Windows`). It's quite obvious, but perhaps this can help a future reader.
```
public partial class MyPlatformBehavior : PlatformBehavior<MyView, object>
{
        protected override void OnAttachedTo(MyView bindable, object platformView)
        {
            // Do your thing
            base.OnAttachedTo(bindable, platformView);
        }

        protected override void OnDetachedFrom(MyView bindable, object platformView)
        {
	    // Do your thing
            base.OnDetachedFrom(bindable, platformView);
        }
}
```
